File: render/figures/score.py
"""The index: the curve, the breakdown, and the weekly panels."""

from __future__ import annotations

import pandas as pd
import plotly.graph_objects as go

from copytext import DOW, t, tpl

from .. import theme
from ..theme import MONO
from .frame import direct_label, frame


def score_line(frames: dict[str, pd.DataFrame], h: int = 340) -> go.Figure:
    fig = go.Figure()
    for user, df in frames.items():
        c = theme.USER_COLOR[user]
        fig.add_trace(go.Scatter(
            x=df["day"], y=df["score"], mode="markers",
            marker=dict(size=4, color=c, opacity=.3),
            name=t("series.daily_plain", user=user), showlegend=False,
            hoverinfo="skip"))
        fig.add_trace(go.Scatter(
            x=df["day"], y=df["score_7d"], mode="lines",
            line=dict(color=c, width=2.6), name=t("series.user", user=user),
            hovertemplate=tpl("hover.score", user=user)))
        direct_label(fig, df["day"].iloc[-1], df["score_7d"].iloc[-1], user, c)
    fig.update_layout(title=t("chart.score_line"),
                      yaxis_title=t("unit.score"))
    fig.update_yaxes(range=[0, 100], dtick=20)
    fig.update_xaxes(tickformat="%d %b")
    return frame(fig, h)


def score_breakdown(contrib: pd.DataFrame, user: str, h: int = 300) -> go.Figure:
    """How many points each component contributes and how many it lets go."""
    d = contrib.iloc[::-1]
    fig = go.Figure()
    fig.add_trace(go.Bar(
        x=d["points"], y=d["component"], orientation="h",
        name=t("series.points_earned"),
        marker=dict(color=theme.USER_COLOR[user], line=dict(color=theme.CARD, width=1.5)),
        hovertemplate=tpl("hover.points_earned"),
        customdata=d["weight"] * 100))
    fig.add_trace(go.Bar(
        x=d["lost"], y=d["component"], orientation="h",
        name=t("series.points_lost"),
        marker=dict(color=theme.LOST, line=dict(color=theme.CARD, width=1.5)),
        hovertemplate=tpl("hover.points_lost")))
    # "Long disconnection" is the longest tick label on the page and it sits
    # in the left margin, so that margin is sized for it rather than the
    # template's 56px, which cut it off on a phone.
    fig.update_layout(title=t("chart.score_breakdown", user=user),
                      barmode="stack", bargap=.3,
                      xaxis_title=t("axis.points"),
                      margin=dict(t=48, r=24, b=68, l=124))
    fig.update_yaxes(tickfont=dict(family=MONO, size=11, color=theme.INK))
    return frame(fig, h)


# ---------------------------------------------------------------------------
# Month walkthrough


def week_evolution(w: pd.DataFrame, col: str, label: str, unit: str,
                   user: str, sel: int, h: int = 260) -> go.Figure:
    """One magnitude week by week, with the selected week highlighted.

    Short weeks carry a mark: a two-day week averaged next to a seven-day one
    reads as a drop that never happened.
    """
    c = theme.USER_COLOR[user]
    labels = [t("label.week_partial" if p else "label.week", week=i)
              for i, p in zip(w.index, w["is_partial"])]
    fig = go.Figure(go.Bar(
        x=labels, y=w[col],
        marker=dict(color=[c if i == sel else theme.DIM for i in w.index],
                    line=dict(color=theme.CARD, width=1.5)),
        text=[f"{v:,.0f}" if abs(v) >= 10 else f"{v:,.1f}" for v in w[col]],
        textposition="outside",
        textfont=dict(family=MONO, size=11, color=theme.INK_2),
        hovertemplate=tpl("hover.week", unit=unit, label=label),
    ))
    fig.update_layout(title=label, yaxis_title=unit, bargap=.35,
                      height=h, showlegend=False,
                      margin=dict(t=44, r=20, b=36, l=54))
    fig.update_yaxes(range=[0, max(w[col].max() * 1.25, 0.1)])
    return fig


def week_days(df: pd.DataFrame, week: int, col: str, label: str, unit: str,
              user: str, h: int = 300) -> go.Figure:
    """The days of the selected week against the mean of previous weeks."""
    cur = df[df["week"] == week]
    prev = df[df["week"] < week]
    ref = prev[col].mean() if len(prev) else None

    fig = go.Figure(go.Bar(
        x=[DOW[d] for d in cur["dow"]], y=cur[col],
        marker=dict(color=theme.USER_COLOR[user], line=dict(color=theme.CARD, width=1.5)),
        name=t("series.week", week=week),
        hovertemplate=tpl("hover.week_day", unit=unit),
    ))
    if ref is not None:
        fig.add_hline(y=ref, line=dict(color=theme.INK, width=1.6, dash="dot"),
                      annotation_text=t("annotation.prev_mean", mean=ref),
                      annotation_position="top left",
                      annotation_font=dict(family=MONO, size=10, color=theme.INK_2))
    if cur[col].abs().max() == 0:
        fig.add_annotation(xref="paper", yref="paper", x=0.5, y=0.5,
                           text=t("annotation.no_activity_week"),
                           showarrow=False,
                           font=dict(family=MONO, size=11, color=theme.MUTED))
        fig.update_yaxes(range=[0, 1], showticklabels=False)
    fig.update_layout(title=label, yaxis_title=unit, bargap=.3, height=h,
                      showlegend=False, margin=dict(t=52, r=20, b=40, l=54))
    return fig


def week_components(w: pd.DataFrame, sel: int, h: int = 320) -> go.Figure:
    """The five index components, week by week."""
    from balance.score import COMPONENTS
    fig = go.Figure()
    labels = [t("label.week", week=i) for i in w.index]
    for (col, label, *_rest), color in zip(COMPONENTS, theme.CATEGORICAL):
        fig.add_trace(go.Scatter(
            x=labels, y=w[f"score_{col}"], mode="lines+markers", name=label,
            line=dict(color=color, width=2.2), marker=dict(size=8, color=color),
            hovertemplate=tpl("hover.component", label=label),
        ))
    fig.add_vline(x=t("label.week", week=sel),
                  line=dict(color=theme.INK, width=1.6, dash="dot"))
    fig.update_layout(title=t("chart.week_components"),
                      yaxis_title=t("unit.score"), height=h,
                      margin=dict(t=44, r=20, b=76, l=54))
    fig.update_yaxes(range=[0, 105], dtick=25)
    return fig

File: web/charts/score.ts
/**
 * The index: the curve, the breakdown, and the weekly panels.
 */

import {
  DOW, annotation, axis as unit, hover, series, title, weekLabel,
} from "../copy/figures";
import { MONO, userColor, type Surface } from "../theme";
import type { DailyRow, UserId, WeeklyRow } from "../types/index";
import { directLabel, frame } from "./frame";
import type { Annotation, Figure, Trace } from "./plotly";
import type { Frames } from "./series";

/** The five index components, their weights, and what each one is called.
 *  Mirrors `COMPONENTS` in `balance/score.py`. */
export const COMPONENTS = [
  { key: "screen_min", label: "Screen time", good: 90, bad: 360, weight: 0.25 },
  { key: "pickups", label: "Fragmentation", good: 15, bad: 60, weight: 0.20 },
  { key: "night_min", label: "Protected night", good: 0, bad: 60, weight: 0.20 },
  { key: "longest_offline_h", label: "Long disconnection", good: 4, bad: 1,
    weight: 0.15 },
  { key: "distract_share", label: "Intent", good: 0.10, bad: 0.50,
    weight: 0.20 },
] as const;

/** Both people's index, every day, with the 7-day mean on top. */
export function scoreLine(s: Surface, frames: Frames): Figure {
  const data: Trace[] = [];
  const labels: Annotation[] = [];
  for (const [user, rows] of Object.entries(frames)) {
    const colour = userColor(s, user);
    const days = rows.map((r) => r.day);
    data.push({
      type: "scatter", mode: "markers", x: days, y: rows.map((r) => r.score),
      marker: { size: 4, color: colour, opacity: 0.3 },
      name: series.dailyPlain(user), showlegend: false, hoverinfo: "skip",
    });
    const smoothed = rows.map((r) => r.score_7d);
    data.push({
      type: "scatter", mode: "lines", x: days, y: smoothed,
      line: { color: colour, width: 2.6 }, name: series.user(user),
      hovertemplate: hover.score(user),
    });
    const lastDay = days[days.length - 1];
    const lastValue = smoothed[smoothed.length - 1];
    if (lastDay !== undefined && lastValue != null) {
      labels.push(directLabel(lastDay, lastValue, user, colour));
    }
  }
  return frame(s, data, {
    title: { text: title.scoreLine },
    yaxis: { title: { text: unit.score }, range: [0, 100], dtick: 20 },
    xaxis: { tickformat: "%d %b" },
    annotations: labels,
  });
}

/** How many points each component contributes on an average day, and how many
 *  it lets go. */
export function scoreBreakdown(s: Surface, rows: DailyRow[],
                               user: UserId): Figure {
  // Plotly draws horizontal bars bottom-up, so the list is reversed to read
  // top-down in the order the components are declared.
  const parts = [...COMPONENTS].reverse().map((c) => {
    const scores = rows.map((r) => r[`score_${c.key}`] as number);
    const mean = scores.reduce((a, b) => a + b, 0) / (scores.length || 1);
    return { ...c, points: mean * c.weight, lost: (100 - mean) * c.weight };
  });
  return frame(s, [
    {
      type: "bar", orientation: "h",
      x: parts.map((p) => p.points), y: parts.map((p) => p.label),
      name: series.pointsEarned,
      marker: { color: userColor(s, user),
                line: { color: s.card, width: 1.5 } },
      customdata: parts.map((p) => p.weight * 100),
      hovertemplate: hover.pointsEarned,
    },
    {
      type: "bar", orientation: "h",
      x: parts.map((p) => p.lost), y: parts.map((p) => p.label),
      name: series.pointsLost,
      marker: { color: s.lost, line: { color: s.card, width: 1.5 } },
      hovertemplate: hover.pointsLost,
    },
  ], {
    title: { text: title.scoreBreakdown(user) },
    xaxis: { title: { text: unit.points } },
    yaxis: { tickfont: { family: MONO, size: 11, color: s.ink } },
    barmode: "stack", bargap: 0.3,
    // "Long disconnection" is the longest tick label on the page and it sits
    // in the left margin, so that margin is sized for it rather than the
    // template's 56, which cut it off on a phone.
    margin: { t: 48, r: 24, b: 68, l: 124 },
  }, 300);
}

// ---------------------------------------------------------------------------
// Month walkthrough
// ---------------------------------------------------------------------------

/** One magnitude week by week, with the selected week highlighted. */
export function weekEvolution(s: Surface, weeks: WeeklyRow[],
                              key: keyof WeeklyRow, heading: string,
                              unitLabel: string, user: UserId,
                              selected: number): Figure {
  const colour = userColor(s, user);
  const values = weeks.map((w) => (w[key] as number | null) ?? 0);
  const most = Math.max(...values, 0);
  return frame(s, [{
    type: "bar",
    x: weeks.map((w) => weekLabel(w.week, w.is_partial)), y: values,
    marker: {
      color: weeks.map((w) => (w.week === selected ? colour : s.dim)),
      line: { color: s.card, width: 1.5 },
    },
    text: values.map((v) => (Math.abs(v) >= 10 ? v.toFixed(0) : v.toFixed(1))),
    textposition: "outside",
    textfont: { family: MONO, size: 11, color: s.ink2 },
    hovertemplate: hover.week(heading, unitLabel),
  }], {
    title: { text: heading },
    yaxis: { title: { text: unitLabel }, range: [0, Math.max(most * 1.25, 0.1)] },
    bargap: 0.35, showlegend: false,
    margin: { t: 44, r: 20, b: 36, l: 54 },
  }, 260);
}

/**
 * The days of the selected week against the mean of the weeks before it.
 *
 * A week is the Nth block of seven days, so week 1 opens on whatever weekday
 * the log does. The bars are still read Monday to Sunday: each block holds one
 * of every weekday, so sorting on it loses nothing.
 */
export function weekDays(s: Surface, rows: DailyRow[], week: number,
                         key: keyof DailyRow, heading: string,
                         unitLabel: string, user: UserId): Figure {
  const current = rows.filter((r) => r.week === week)
                      .sort((a, b) => a.dow - b.dow);
  const earlier = rows.filter((r) => r.week < week);
  const values = current.map((r) => (r[key] as number | null) ?? 0);

  const shapes = [];
  const annotations: Annotation[] = [];
  if (earlier.length) {
    const mean = earlier.reduce((sum, r) => sum + ((r[key] as number) ?? 0), 0)
                 / earlier.length;
    shapes.push({ type: "line" as const, xref: "paper", x0: 0, x1: 1,
                  y0: mean, y1: mean,
                  line: { color: s.ink, width: 1.6, dash: "dot" } });
    annotations.push({ xref: "paper", x: 0, y: mean,
                       text: annotation.prevMean(mean), showarrow: false,
                       xanchor: "left", yanchor: "bottom", yshift: 3,
                       font: { family: MONO, size: 10, color: s.ink2 } });
  }
  if (Math.max(...values.map(Math.abs), 0) === 0) {
    annotations.push({ xref: "paper", yref: "paper", x: 0.5, y: 0.5,
                       text: annotation.noActivityWeek, showarrow: false,
                       font: { family: MONO, size: 11, color: s.muted } });
  }

  return frame(s, [{
    type: "bar",
    x: current.map((r) => DOW[r.dow] ?? String(r.dow)), y: values,
    marker: { color: userColor(s, user),
              line: { color: s.card, width: 1.5 } },
    name: series.week(week),
    hovertemplate: hover.weekDay(unitLabel),
  }], {
    title: { text: heading },
    yaxis: { title: { text: unitLabel } },
    xaxis: { categoryorder: "array", categoryarray: [...DOW] },
    bargap: 0.3, showlegend: false,
    margin: { t: 52, r: 20, b: 40, l: 54 },
    shapes, annotations,
  }, 300);
}

/** The five index components, scored 0 to 100, week by week. */
export function weekComponents(s: Surface, weeks: WeeklyRow[],
                               selected: number): Figure {
  const labels = weeks.map((w) => weekLabel(w.week));
  const data: Trace[] = COMPONENTS.map((component, i) => {
    const colour = s.categorical[i % s.categorical.length] ?? s.ink;
    return {
      type: "scatter", mode: "lines+markers", x: labels,
      y: weeks.map((w) => w[`score_${component.key}`] as number),
      name: component.label,
      line: { color: colour, width: 2.2 },
      marker: { size: 8, color: colour },
      hovertemplate: hover.component(component.label),
    };
  });
  return frame(s, data, {
    title: { text: title.weekComponents },
    yaxis: { title: { text: unit.score }, range: [0, 105], dtick: 25 },
    shapes: [{ type: "line", yref: "paper", y0: 0, y1: 1,
               x0: weekLabel(selected), x1: weekLabel(selected),
               line: { color: s.ink, width: 1.6, dash: "dot" } }],
    margin: { t: 44, r: 20, b: 76, l: 54 },
  }, 320);
}
